### PR TITLE
Typing for accelerators and plugins

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Part of #7035 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [n/a] Did you write any new necessary tests? (not for typos and docs)
- [n/a] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/accelerators/tpu.py [hunk 1]
```diff
@@ -51,27 +50,24 @@ def setup(self, trainer: 'pl.Trainer', model: 'pl.LightningModule') -> None:
...
 
-        model = self.lightning_module
-        parameters = model.parameters()
-
         grad_clip_val = float(clip_val)
         if grad_clip_val <= 0:
             return
 
-        max_norm = grad_clip_val
+        parameters = self.model.parameters()
```
Comment (carmocca) [inline pytorch_lightning/accelerators/tpu.py:70]: Is this equivalent to what we had before?

File: pytorch_lightning/accelerators/tpu.py [hunk 2]
```diff
@@ -51,27 +50,24 @@ def setup(self, trainer: 'pl.Trainer', model: 'pl.LightningModule') -> None:
...
+        norm_type = 2.0
 
-        xla_clip_grad_norm_(parameters, max_norm, norm_type)
+        clip_grad_norm_(parameters, grad_clip_val, norm_type)
```
Comment (kaushikb11) [inline pytorch_lightning/accelerators/tpu.py:73]: @carmocca Don't change this. I had this for https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/models/test_tpu.py#L386

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/plugins/precision/precision_plugin.py [hunk 3]
```diff
@@ -119,24 +114,25 @@ def clip_gradients(
         if clip_val <= 0:
             return
 
-        clip_grad_func = self.clip_grad_funcs[gradient_clip_algorithm]
-        clip_grad_func(optimizer, clip_val)  # type: ignore
+        if gradient_clip_algorithm == GradClipAlgorithmType.VALUE:
+            self.clip_grad_by_value(optimizer, clip_val)
+        elif gradient_clip_algorithm == GradClipAlgorithmType.NORM:
+            # TODO: there should be a mechanism to set `norm_type`
+            self.clip_grad_by_norm(optimizer, clip_val, eps=self.EPSILON)
```
Comment (awaelchli) [inline pytorch_lightning/plugins/precision/precision_plugin.py:121]: this is different here, are you sure?

File: pytorch_lightning/accelerators/accelerator.py [hunk 4]
```diff
@@ -105,28 +108,28 @@ def pre_dispatch(self, trainer: 'pl.Trainer') -> None:
             self.setup_optimizers(trainer)
         self.precision_plugin.pre_dispatch()
 
-    def post_dispatch(self, trainer: 'pl.Trainer') -> None:
+    def post_dispatch(self, _: 'pl.Trainer') -> None:
```
Comment (awaelchli) [inline pytorch_lightning/accelerators/accelerator.py:111]: is this needed?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/accelerators/accelerator.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/accelerators/accelerator.py:111]: Can be undone

```suggestion
    def post_dispatch(self, trainer: 'pl.Trainer') -> None:
```

File: pytorch_lightning/plugins/precision/precision_plugin.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/plugins/precision/precision_plugin.py:121]: I couldn't get mypy to work with the previous way of pulling the function out of the dict.

I think having `if-elif` is simpler and okay

File: pytorch_lightning/accelerators/tpu.py [hunk 5]
```diff
@@ -59,19 +59,16 @@ def clip_gradients(
         self,
         optimizer: Optimizer,
         clip_val: Union[float, int],
-        norm_type: float = 2.0,
```
Comment (Borda) [inline pytorch_lightning/accelerators/tpu.py:62]: this removed on purpose?

File: pytorch_lightning/accelerators/tpu.py [hunk 6]
```diff
@@ -59,19 +59,16 @@ def clip_gradients(
...
-        model = self.lightning_module
-        parameters = model.parameters()
-
         grad_clip_val = float(clip_val)
         if grad_clip_val <= 0:
             return
 
-        max_norm = grad_clip_val
+        parameters = self.model.parameters()
+        norm_type = 2.0
```
Comment (Borda) [inline pytorch_lightning/accelerators/tpu.py:72]: how about this hardcoded 2?

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 7]
```diff
@@ -49,28 +57,29 @@ def connect(self, model: torch.nn.Module, optimizers: Sequence[Optimizer],
...
             optimizer: the optimizer to perform the step lateron
-            opt_idx: the optimizer's index
+            opt_idx: the optimizer index
             should_accumulate: whether to accumulate gradients or not
 
         """
-        closure_loss = amp.scale_loss(closure_loss, model.trainer.optimizers if optimizer is None else optimizer)
+        scaled_loss: ContextManager[Tensor] = amp.scale_loss(
+            closure_loss, model.trainer.optimizers if optimizer is None else optimizer
+        )
```
Comment (Borda) [inline pytorch_lightning/plugins/precision/apex_amp.py:79]:
```suggestion
        opt = model.trainer.optimizers if optimizer is None else optimizer
        scaled_loss: ContextManager[Tensor] = amp.scale_loss(closure_loss, opt)
```

File: pytorch_lightning/plugins/precision/precision_plugin.py [hunk 8]
```diff
@@ -32,41 +35,34 @@ class PrecisionPlugin(Plugin):
     EPSILON: float = 1e-6
     precision: Union[str, int] = 32
 
-    def __init__(self) -> None:
```
Comment (Borda) [inline pytorch_lightning/plugins/precision/precision_plugin.py:35]: no init needed anymore?

Comment (Borda) [reply pytorch_lightning/plugins/precision/precision_plugin.py:35]: I see some related edit bellow, but still, can you comment on it?

File: pytorch_lightning/accelerators/tpu.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/accelerators/tpu.py:62]: Yes, the parent's signature does not have this argument so mypy complains

Also, there is currently no way for users to change the `norm_type` value so it's okay to remove

File: pytorch_lightning/accelerators/tpu.py (hunk 6 above)
Comment (carmocca) [reply pytorch_lightning/accelerators/tpu.py:72]: See my previous comment

File: pytorch_lightning/plugins/precision/precision_plugin.py (hunk 8 above)
Comment (carmocca) [reply pytorch_lightning/plugins/precision/precision_plugin.py:35]: Not needed, it would be empty

**Review by Borda — APPROVED** [review]
